chips/rp2040/src/pwm.rs

```diff
@@ -0,0 +1,1354 @@
+        }
+    }
+}
+
+/// PWM channel configuration structure
+///
+/// This helper struct allows multiple channels to share the same configuration.
+pub struct PwmChannelConfiguration {
+    en: bool,
+    ph_correct: bool,
+    a_inv: bool,
+    b_inv: bool,
+    divmode: DivMode,
+    int: u8,
+    frac: u8,
+    cc_a: u16,
+    cc_b: u16,
+    top: u16,
+}
+
+impl PwmChannelConfiguration {
+    /// Create a set of default values to use for configuring a PWM channel:
+    /// + the channel is disabled
+    /// + trailing-edge modulation configured
+    /// + no pin A and B polarity inversion
+    /// + free running mode for the fractional clock divider
+    /// + integral part of the divider is 1 and the fractional part is 0
+    /// + compare values for both pins are set 0 (0% duty cycle)
+    /// + top value is set to its maximum value
+    pub fn default_config() -> Self {
+        PwmChannelConfiguration {
+            en: false,
+            ph_correct: false,
+            a_inv: false,
+            b_inv: false,
+            divmode: DivMode::FreeRunning,
+            int: 1,
+            frac: 0,
+            cc_a: 0,
+            cc_b: 0,
+            top: u16::MAX,
+        }
+    }
+
+    /// See [Pwm::set_enabled]
+    pub fn set_enabled(&mut self, enable: bool) {
+        self.en = enable;
+    }
+
+    /// See [Pwm::set_ph_correct]
+    pub fn set_ph_correct(&mut self, ph_correct: bool) {
+        self.ph_correct = ph_correct;
+    }
+
+    /// See [Pwm::set_invert_polarity]
+    pub fn set_invert_polarity(&mut self, a_inv: bool, b_inv: bool) {
+        self.a_inv = a_inv;
+        self.b_inv = b_inv;
+    }
+
+    /// See [Pwm::set_div_mode]
+    pub fn set_div_mode(&mut self, divmode: DivMode) {
+        self.divmode = divmode;
+    }
+
+    /// See [Pwm::set_divider_int_frac]
+    pub fn set_divider_int_frac(&mut self, int: u8, frac: u8) {
+        if int == 0 || frac > 15 {
+            return;
+        }
+        self.int = int;
+        self.frac = frac;
+    }
+
+    /// See [Pwm::set_compare_value_a]
+    pub fn set_compare_value_a(&mut self, cc_a: u16) {
+        self.cc_a = cc_a;
+    }
+
+    /// See [Pwm::set_compare_value_b]
+    pub fn set_compare_value_b(&mut self, cc_b: u16) {
+        self.cc_b = cc_b;
+    }
+
+    /// See [Pwm::set_compare_values_a_and_b]
+    pub fn set_compare_values_a_and_b(&mut self, cc_a: u16, cc_b: u16) {
+        self.set_compare_value_a(cc_a);
+        self.set_compare_value_b(cc_b);
+    }
+
+    /// See [Pwm::set_top]
+    pub fn set_top(&mut self, top: u16) {
+        self.top = top;
+    }
+}
+
+const PWM_BASE: StaticRef<PwmRegisters> =
+    unsafe { StaticRef::new(0x40050000 as *const PwmRegisters) };
+
+/// Main struct for controlling PWM peripheral
+pub struct Pwm<'a> {
+    registers: StaticRef<PwmRegisters>,
+    clocks: OptionalCell<&'a clocks::Clocks>,
+    interrupt_handler: OptionalCell<&'a dyn Interrupt>,
+}
+
+impl<'a> Pwm<'a> {
```

Comment:
Methods in this implementation, except `new` should be private, as they should never be used outside the `rp2040`create. Anyone having a reference to this structure should use it via the `Pwm` HIL.

Reply:
I kept the other methods public, in the case someone would need much finer control of the PWM peripheral, like the integration examples show. Of course, one should use the HIL if highly portable code is the priority and if the HIL suffices. As the PWM HIL evolves, more and more methods will become private.

If that is not the intended case, I can make all the methods private with the downside of having only one integration example working and severely limiting the functionalities of the driver.

There is also an `Interrupt` trait that I have created and implemented myself. This will surely be integrated in the HIL at some point and removed from the PWM driver.

Reply:
Made all methods private + removed `PwmChannelConfiguration` methods (used for configuration outside of the HIL).